src/test/java/org/javarosa/core/model/DynamicSelectUpdateTest.java

```diff
@@ -0,0 +1,323 @@
+        Scenario scenario = Scenario.init("Select from repeat", getSelectFromRepeatForm());
+
+        scenario.answer("/data/repeat[0]/value", "a");
+        scenario.answer("/data/repeat[0]/label", "A");
+        assertThat(scenario.choicesOf("/data/select"), contains(
+            choice("a", "A")));
+
+        scenario.removeRepeat("/data/repeat[0]");
+        assertThat(scenario.choicesOf("/data/select").size(), is(0));
+    }
+
+    @Test
+    public void selectFromRepeat_withPredicate_whenPredicateTriggerChanges_updatesChoices() throws Exception {
+        Scenario scenario = Scenario.init("Select from repeat", getSelectFromRepeatForm("starts-with(value,current()/../filter)"));
+
+        scenario.answer("/data/repeat[0]/value", "a");
+        scenario.answer("/data/repeat[0]/label", "A");
+        scenario.answer("/data/filter", "a");
+        assertThat(scenario.choicesOf("/data/select"), contains(
+            choice("a", "A")));
+
+        scenario.answer("/data/filter", "b");
+        assertThat(scenario.choicesOf("/data/select").size(), is(0));
+    }
+
+    private static XFormsElement getSelectFromRepeatForm() {
+        return getSelectFromRepeatForm("");
+    }
+
+    private static XFormsElement getSelectFromRepeatForm(String predicate) {
+        return html(
+            head(
+                title("Select from repeat"),
+                model(
+                    mainInstance(
+                        t("data id=\"repeat-select\"",
+                            t("repeat",
+                                t("value"),
+                                t("label")),
+                            t("select"),
+                            t("filter"))))),
+            body(
+                repeat("/data/repeat",
+                    input("value"),
+                    input("label")),
+                select1Dynamic("/data/select", "../repeat" + (!predicate.isEmpty() ? "[" + predicate + "]" : "")),
+                input("filter")
+            )
+        );
+    }
+    //endregion
+
+    //region Select in repeat
+    // When a dynamic select is in a repeat, the itemsets for all repeat instances are represented by the same ItemsetBinding.
+    @Test
+    public void selectInRepeat_withPredicate_evaluatesChoiceListForEachRepeatInstance() throws Exception {
+        Scenario scenario = Scenario.init("Select from repeat", html(
+            head(
+                title("Select in repeat"),
+                model(
+                    mainInstance(
+                        t("data id=\"repeat-select\"",
+                            t("repeat",
+                                t("select"),
+                                t("filter")))),
+
+                    instance("choices",
+                        item("a", "A"),
+                        item("aa", "AA"),
+                        item("b", "B"),
+                        item("bb", "BB"))),
+                body(
```

Comment:
Is it OK for the body to be inside the head?

Reply:
<img width="1855" alt="image" src="https://user-images.githubusercontent.com/55099/109876528-c1fb3780-7c26-11eb-99a5-11f2e801643d.png">

If it is invalid, then perhaps we’d modify the DSL to disallow that at compile time.

Reply:
Huh. I'm shocked that works. It's pretty twisted.

Looks like we introduced that badness in a test at some point and then kept copying it everywhere. Good catch. Let me know if there's a less verbose way to introduce `Head/BodyXFormsElement`.

Reply:
Maybe this?

<img width="1454" alt="image" src="https://user-images.githubusercontent.com/55099/109911560-2934de00-7c5f-11eb-8ce0-7208711928b7.png">

```
    class HeadXFormsElement extends TagXFormsElement {
        public HeadXFormsElement(XFormsElement[] children) {
            super("h:head", emptyMap(), Arrays.asList(children));
        }
    }

    class BodyXFormsElement extends TagXFormsElement {
        public BodyXFormsElement(XFormsElement[] children) {
            super("h:body", emptyMap(), Arrays.asList(children));
        }
    }
```

Reply:
🧐 🧐 🧐 There was a reason I thought extending `TagXFormsElement` wasn't what I wanted at the time but now I have no idea what it might have been.